src/data/value/mod.rs

```diff
@@ -315,6 +315,33 @@ impl Value {
         }
     }
 
+    pub fn unary_factorial(&self) -> Result<Value> {
+        use Value::*;
+
+        let factorial_function = |a: i64| -> Result<i64> {
+            (1..(a + 1))
+                .into_iter()
+                .try_fold(1i64, |mul, x| mul.checked_mul(x))
+                .ok_or_else(|| ValueError::FactorialOverflow.into())
+        };
+
+        match self {
+            I64(a) => {
+                if *a < 0 {
+                    Err(ValueError::FactorialOnNegativeNumeric.into())
+                } else {
+                    match factorial_function(*a) {
+                        Ok(x) => Ok(I64(x)),
+                        Err(x) => Err(x),
+                    }
+                }
+            }
+            F64(_) => Err(ValueError::FactorialOnNonInteger.into()),
+            Null => Ok(Null),
+            _ => Err(ValueError::FactorialOnNonNumeric.into()),
+        }
```

Comment:
Tiny simplification, Rust `match` is quite powerful.
```rust
match self {                                                     
    I64(a) if *a >= 0 => factorial_function(*a).map(I64),        
    I64(_) => Err(ValueError::FactorialOnNegativeNumeric.into()),
    F64(_) => Err(ValueError::FactorialOnNonInteger.into()),     
    Null => Ok(Null),                                            
    _ => Err(ValueError::FactorialOnNonNumeric.into()),          
}                                                                
```

Reply:
@panarch 

thanks.
aligned line..

Reply:
@panarch

change code..